File: sdk/trustpulse.py
"""
TrustPulse Python SDK
─────────────────────
Drop-in client for server-side integrations (Django, Flask, FastAPI middleware).

pip install trustpulse-sdk   # or: pip install httpx

Usage
─────
from trustpulse import TrustPulseClient, Event

client = TrustPulseClient(endpoint="http://localhost:8000/v1")

result = client.score(
    user_id="user_123",
    device_id="dev_abc",
    event=Event.TRANSFER,
    transaction_amount=85000,
    is_known_device=True,
)

if result.action == "silent_pass":
    pass   # proceed transparently
elif result.action == "biometric_prompt":
    trigger_biometric_flow()
elif result.action == "step_up_auth":
    trigger_otp_flow()
else:  # block_and_alert
    raise PermissionError("Session blocked by TrustPulse.")
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional

try:
    import httpx
    _HAS_HTTPX = True
except ImportError:
    import urllib.request, json as _json
    _HAS_HTTPX = False


# ─── Enumerations ─────────────────────────────────────────────────────────────

class Event(str, Enum):
    LOGIN             = "login"
    TRANSFER          = "transfer"
    BENEFICIARY_ADD   = "beneficiary_add"
    ACCOUNT_RECOVERY  = "account_recovery"
    ADMIN_ACCESS      = "admin_access"
    SIM_ACTIVITY      = "sim_activity"


class NetworkType(str, Enum):
    WIFI    = "wifi"
    G4      = "4g"
    G5      = "5g"
    G3      = "3g"
    VPN     = "vpn"
    TOR     = "tor"
    UNKNOWN = "unknown"


class Action(str, Enum):
    SILENT_PASS      = "silent_pass"
    BIOMETRIC_PROMPT = "biometric_prompt"
    STEP_UP_AUTH     = "step_up_auth"
    BLOCK_AND_ALERT  = "block_and_alert"


# ─── Data Classes ─────────────────────────────────────────────────────────────

@dataclass
class ScoreRequest:
    user_id:                    str
    device_id:                  str
    event:                      Event           = Event.LOGIN
    session_id:                 Optional[str]   = None

    # Device trust
    is_known_device:            bool            = True
    device_health_score:        float           = 1.0
    vpn_detected:               bool            = False
    network_type:               NetworkType     = NetworkType.WIFI

    # Behavioural (on-device derived scores only)
    keystroke_anomaly:          float           = 0.0
    touch_anomaly:              float           = 0.0
    gait_anomaly:               float           = 0.0

    # Context
    location_anomaly:           float           = 0.0
    time_anomaly:               float           = 0.0

    # Event semantics
    transaction_amount:         Optional[float] = None
    is_new_beneficiary:         bool            = False
    failed_attempts_last_hour:  int             = 0

    def __post_init__(self):
        if self.session_id is None:
            self.session_id = f"sess_{uuid.uuid4().hex[:12]}"

    def to_dict(self) -> Dict:
        return {
            "session_id":                  self.session_id,
            "user_id":                     self.user_id,
            "device_id":                   self.device_id,
            "event_type":                  self.event.value,
            "is_known_device":             self.is_known_device,
            "device_health_score":         self.device_health_score,
            "vpn_detected":                self.vpn_detected,
            "network_type":                self.network_type.value,
            "keystroke_anomaly":           self.keystroke_anomaly,
            "touch_anomaly":               self.touch_anomaly,
            "gait_anomaly":                self.gait_anomaly,
            "location_anomaly":            self.location_anomaly,
            "time_anomaly":                self.time_anomaly,
            "transaction_amount":          self.transaction_amount,
            "is_new_beneficiary":          self.is_new_beneficiary,
            "failed_attempts_last_hour":   self.failed_attempts_last_hour,
        }


@dataclass
class ScoreResult:
    session_id:      str
    trust_score:     int
    risk_level:      str
    action:          Action
    latency_ms:      float
    explanation:     List[str]
    breakdown:       Dict[str, int]
    audit_ref:       str

    @classmethod
    def from_dict(cls, data: Dict) -> "ScoreResult":
        return cls(
            session_id=  data["session_id"],
            trust_score= data["trust_score"],
            risk_level=  data["risk_level"],
            action=      Action(data["required_action"]),
            latency_ms=  data.get("latency_ms", 0.0),
            explanation= data.get("explanation", []),
            breakdown=   data.get("breakdown", {}),
            audit_ref=   data.get("audit_ref", ""),
        )

    @property
    def is_blocked(self) -> bool:
        return self.action == Action.BLOCK_AND_ALERT

    @property
    def needs_verification(self) -> bool:
        return self.action in (
            Action.BIOMETRIC_PROMPT, Action.STEP_UP_AUTH
        )


# ─── Client ───────────────────────────────────────────────────────────────────

class TrustPulseClient:
    """
    Synchronous TrustPulse API client.

    Args:
        endpoint:  TrustPulse API base URL (e.g. 'https://api.trustpulse.io/v1')
        api_key:   API key for authenticated requests
        timeout:   Request timeout in seconds (default: 2.0)
        fail_open: If True, return 'silent_pass' on API errors instead of raising.
                   Recommended for production to avoid auth outages.
    """

    def __init__(
        self,
        endpoint: str = "http://localhost:8000/v1",
        api_key: Optional[str] = None,
        timeout: float = 2.0,
        fail_open: bool = True,
    ):
        self.endpoint  = endpoint.rstrip("/")
        self.api_key   = api_key
        self.timeout   = timeout
        self.fail_open = fail_open
        self._headers  = {
            "Content-Type": "application/json",
            **({"X-API-Key": api_key} if api_key else {}),
        }

    def score(
        self,
        user_id: str,
        device_id: str,
        event: Event = Event.LOGIN,
        **kwargs,
    ) -> ScoreResult:
        """
        Compute a trust score.

        Returns a ScoreResult. Check .action to decide the authentication path.
        On network errors, behaviour depends on fail_open setting.
        """
        req = ScoreRequest(
            user_id=user_id,
            device_id=device_id,
            event=event,
            **kwargs,
        )
        return self._post(req)

    def _post(self, req: ScoreRequest) -> ScoreResult:
        url     = f"{self.endpoint}/score"
        payload = req.to_dict()

        try:
            if _HAS_HTTPX:
                with httpx.Client(timeout=self.timeout) as client:
                    resp = client.post(url, json=payload, headers=self._headers)
                    resp.raise_for_status()
                    return ScoreResult.from_dict(resp.json())
            else:
                import json
                data = json.dumps(payload).encode()
                request = urllib.request.Request(
                    url, data=data,
                    headers={**self._headers, "Content-Length": str(len(data))},
                    method="POST",
                )
                with urllib.request.urlopen(request, timeout=self.timeout) as r:
                    return ScoreResult.from_dict(json.loads(r.read()))

        except Exception as exc:
            if self.fail_open:
                # Fail open: allow the session through rather than block all users
                return ScoreResult(
                    session_id=  req.session_id or "",
                    trust_score= 0,
                    risk_level=  "low",
                    action=      Action.SILENT_PASS,
                    latency_ms=  0.0,
                    explanation= [f"Fallback: API unreachable ({exc})"],
                    breakdown=   {},
                    audit_ref=   "fallback",
                )
            raise


class AsyncTrustPulseClient(TrustPulseClient):
    """Async version for use with FastAPI / asyncio."""

    async def score_async(
        self,
        user_id: str,
        device_id: str,
        event: Event = Event.LOGIN,
        **kwargs,
    ) -> ScoreResult:
        import httpx as _httpx
        req     = ScoreRequest(user_id=user_id, device_id=device_id, event=event, **kwargs)
        url     = f"{self.endpoint}/score"
        payload = req.to_dict()

        try:
            async with _httpx.AsyncClient(timeout=self.timeout) as client:
                resp = await client.post(url, json=payload, headers=self._headers)
                resp.raise_for_status()
                return ScoreResult.from_dict(resp.json())
        except Exception as exc:
            if self.fail_open:
                return ScoreResult(
                    session_id=req.session_id or "", trust_score=0,
                    risk_level="low", action=Action.SILENT_PASS,
                    latency_ms=0.0, explanation=[f"Fallback: {exc}"],
                    breakdown={}, audit_ref="fallback",
                )
            raise


# ─── Django middleware example ────────────────────────────────────────────────

DJANGO_MIDDLEWARE_EXAMPLE = """
# settings.py
TRUSTPULSE_ENDPOINT = "https://api.trustpulse.io/v1"
TRUSTPULSE_API_KEY  = os.environ["TRUSTPULSE_KEY"]
TRUSTPULSE_FAIL_OPEN = True

# middleware.py
from trustpulse import TrustPulseClient, Event
from django.conf import settings

_tp = TrustPulseClient(
    endpoint=settings.TRUSTPULSE_ENDPOINT,
    api_key=settings.TRUSTPULSE_API_KEY,
    fail_open=settings.TRUSTPULSE_FAIL_OPEN,
)

class TrustPulseMiddleware:
    PROTECTED_PATHS = {
        '/api/transfer/':          Event.TRANSFER,
        '/api/beneficiary/add/':   Event.BENEFICIARY_ADD,
        '/api/account/recover/':   Event.ACCOUNT_RECOVERY,
        '/admin/':                 Event.ADMIN_ACCESS,
    }

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        event = next(
            (e for p, e in self.PROTECTED_PATHS.items() if request.path.startswith(p)),
            None
        )
        if event and request.user.is_authenticated:
            result = _tp.score(
                user_id=str(request.user.id),
                device_id=request.META.get('HTTP_X_DEVICE_ID', 'unknown'),
                event=event,
            )
            request.trust_result = result
            if result.is_blocked:
                from django.http import JsonResponse
                return JsonResponse(
                    {"error": "Session blocked", "audit_ref": result.audit_ref},
                    status=403,
                )
        return self.get_response(request)
"""

File: sdk/trustpulse.js
/**
 * TrustPulse SDK — Zero-Friction Adaptive Identity Trust Engine
 * BOB Hackathon 2026 | IIT Gandhinagar
 *
 * Privacy contract
 * ─────────────────
 * All behavioural signals are processed ON-DEVICE using TF.js / native inference.
 * Only the derived anomaly SCORE (a single float 0.0–1.0) is transmitted.
 * No raw biometric data, keystrokes, touch coordinates, or sensor readings
 * ever leave the device. DPDP Act 2023 compliant.
 *
 * Usage (3 lines)
 * ────────────────
 *   import { TrustPulse } from './trustpulse.js';
 *
 *   const tp = new TrustPulse({ endpoint: 'https://api.trustpulse.io/v1' });
 *
 *   const result = await tp.score({ eventType: 'transfer', amount: 50000 });
 *   // result.required_action → 'silent_pass' | 'biometric_prompt' |
 *   //                           'step_up_auth' | 'block_and_alert'
 */

export class TrustPulse {
  /**
   * @param {object} options
   * @param {string}   options.endpoint    - TrustPulse API base URL
   * @param {string}   [options.apiKey]    - API key (omit in dev/sandbox)
   * @param {Function} [options.onDecision] - Callback fired after every score
   * @param {boolean}  [options.debug]     - Log to console
   */
  constructor({ endpoint, apiKey, onDecision, debug = false } = {}) {
    this.endpoint   = (endpoint || 'http://localhost:8000/v1').replace(/\/$/, '');
    this.apiKey     = apiKey;
    this.onDecision = onDecision;
    this.debug      = debug;

    this._sessionId      = this._mkId('sess');
    this._deviceCache    = null;
    this._behavior       = new BehaviorCollector({ debug });

    this._behavior.start();
    if (debug) console.log('[TrustPulse] SDK initialised. Session:', this._sessionId);
  }

  // ─────────────────────────────────────────────────────────────────
  //  Public API
  // ─────────────────────────────────────────────────────────────────

  /**
   * Compute a trust score and return an authentication decision.
   *
   * @param {object} ctx - Event context
   * @param {string}  ctx.eventType       - 'login' | 'transfer' | 'beneficiary_add' |
   *                                        'account_recovery' | 'admin_access' | 'sim_activity'
   * @param {string}  [ctx.userId]
   * @param {number}  [ctx.amount]        - Transaction amount in INR
   * @param {boolean} [ctx.isNewBeneficiary]
   * @param {number}  [ctx.failedAttempts]
   * @returns {Promise<ScoreResult>}
   */
  async score(ctx = {}) {
    const [device, behavior, location, timing] = await Promise.all([
      this._getDeviceSignals(),
      this._behavior.getSignals(),
      this._getLocationAnomaly(),
      Promise.resolve(this._getTimeAnomaly()),
    ]);

    const payload = {
      session_id:               this._sessionId,
      user_id:                  ctx.userId || 'anonymous',

      // Device
      device_id:                device.id,
      is_known_device:          device.isKnown,
      device_health_score:      device.healthScore,
      vpn_detected:             device.vpnDetected,
      network_type:             device.networkType,

      // Behaviour (on-device scores only — no raw data)
      keystroke_anomaly:        behavior.keystrokeAnomaly,
      touch_anomaly:            behavior.touchAnomaly,
      gait_anomaly:             behavior.gaitAnomaly,

      // Context
      location_anomaly:         location,
      time_anomaly:             timing,

      // Event
      event_type:               ctx.eventType          || 'login',
      transaction_amount:       ctx.amount             || null,
      is_new_beneficiary:       ctx.isNewBeneficiary   || false,
      failed_attempts_last_hour: ctx.failedAttempts    || 0,
    };

    if (this.debug) console.log('[TrustPulse] Scoring payload:', payload);

    const headers = { 'Content-Type': 'application/json' };
    if (this.apiKey) headers['X-API-Key'] = this.apiKey;

    const res = await fetch(`${this.endpoint}/score`, {
      method:  'POST',
      headers,
      body:    JSON.stringify(payload),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`TrustPulse API error ${res.status}: ${err}`);
    }

    /** @type {ScoreResult} */
    const result = await res.json();

    if (this.debug) console.log('[TrustPulse] Result:', result);
    if (this.onDecision) this.onDecision(result);

    return result;
  }

  /**
   * Mark the current device as trusted for this user.
   * Call this after a successful high-friction authentication.
   */
  trustDevice() {
    const id = this._deviceCache?.id;
    if (!id) return;
    const known = JSON.parse(localStorage.getItem('_tp_kd') || '[]');
    if (!known.includes(id)) {
      known.push(id);
      localStorage.setItem('_tp_kd', JSON.stringify(known.slice(-20)));
    }
  }

  /** Stop behaviour collection. Call on logout. */
  destroy() {
    this._behavior.stop();
    if (this.debug) console.log('[TrustPulse] SDK destroyed.');
  }

  // ─────────────────────────────────────────────────────────────────
  //  Internal signal collectors
  // ─────────────────────────────────────────────────────────────────

  async _getDeviceSignals() {
    if (this._deviceCache) return this._deviceCache;

    const id          = await this._fingerprint();
    const knownIds    = JSON.parse(localStorage.getItem('_tp_kd') || '[]');

    this._deviceCache = {
      id,
      isKnown:      knownIds.includes(id),
      healthScore:  this._deviceHealth(),
      vpnDetected:  await this._detectVPN(),
      networkType:  this._networkType(),
    };
    return this._deviceCache;
  }

  /** Lightweight, privacy-safe device fingerprint using public browser attributes. */
  async _fingerprint() {
    const parts = [
      navigator.userAgent,
      navigator.language,
      `${screen.width}x${screen.height}x${screen.colorDepth}`,
      Intl.DateTimeFormat().resolvedOptions().timeZone,
      String(navigator.hardwareConcurrency || 0),
      String(navigator.deviceMemory       || 0),
    ].join('|');

    const buf  = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(parts));
    const hex  = Array.from(new Uint8Array(buf))
                      .map(b => b.toString(16).padStart(2, '0'))
                      .join('');
    return hex.slice(0, 20);
  }

  _deviceHealth() {
    let score = 1.0;
    if (!window.isSecureContext)    score -= 0.4;   // Not HTTPS
    if (!navigator.cookieEnabled)   score -= 0.1;
    return Math.max(0, Math.round(score * 100) / 100);
  }

  /** Heuristic VPN detection: compares browser timezone with reported locale.
   *  In production, cross-check against IP geolocation service. */
  async _detectVPN() {
    try {
      const tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
      // Flag common anonymising timezones if they don't match locale
      return tz === 'UTC' && navigator.language !== 'en-US';
    } catch {
      return false;
    }
  }

  _networkType() {
    const conn = navigator.connection || navigator.mozConnection || navigator.webkitConnection;
    return conn?.effectiveType || 'unknown';
  }

  /** Compare current IP geolocation against stored historical centroid.
   *  Stub — implement via IP geolocation API in production. */
  async _getLocationAnomaly() {
    return 0.0;
  }

  /** Score time-of-day anomaly based on UTC hour. */
  _getTimeAnomaly() {
    const h = new Date().getUTCHours();
    if (h >= 1 && h <= 5) return 0.85;   // 1 AM – 5 AM UTC
    if (h === 0 || h === 6) return 0.4;
    return 0.0;
  }

  _mkId(prefix) {
    return `${prefix}_${crypto.randomUUID().replace(/-/g, '').slice(0, 14)}`;
  }
}


// ─────────────────────────────────────────────────────────────────────────────
//  Behaviour Collector
//  Collects raw interaction events ON-DEVICE and derives anomaly scores.
//  The raw events are never exposed outside this class.
// ─────────────────────────────────────────────────────────────────────────────

class BehaviorCollector {
  constructor({ debug = false } = {}) {
    this.debug   = debug;
    this._active = false;

    // Raw event buffers — private, never serialised
    this._keyIntervals  = [];   // ms between keydown events
    this._touchPressure = [];   // touch force values (0–1)
    this._scrollRates   = [];   // px/s scroll velocity

    this._lastKeyTime  = null;
    this._lastScrollY  = 0;
    this._lastScrollTs = null;

    // Bound handlers so we can remove them on stop()
    this._onKey    = this._handleKey.bind(this);
    this._onTouch  = this._handleTouch.bind(this);
    this._onScroll = this._handleScroll.bind(this);
  }

  start() {
    if (this._active) return;
    this._active = true;
    window.addEventListener('keydown',    this._onKey);
    window.addEventListener('touchstart', this._onTouch, { passive: true });
    window.addEventListener('scroll',     this._onScroll, { passive: true });
  }

  stop() {
    this._active = false;
    window.removeEventListener('keydown',    this._onKey);
    window.removeEventListener('touchstart', this._onTouch);
    window.removeEventListener('scroll',     this._onScroll);
  }

  /** Returns on-device derived anomaly scores (0.0–1.0). No raw data. */
  getSignals() {
    return {
      keystrokeAnomaly: this._keystrokeAnomaly(),
      touchAnomaly:     this._touchAnomaly(),
      gaitAnomaly:      0.0,   // Requires native accelerometer via Capacitor/RN
    };
  }

  // ── Private collectors ────────────────────────────────────────────

  _handleKey(e) {
    const now = Date.now();
    if (this._lastKeyTime !== null) {
      this._keyIntervals.push(now - this._lastKeyTime);
      if (this._keyIntervals.length > 60) this._keyIntervals.shift();
    }
    this._lastKeyTime = now;
  }

  _handleTouch(e) {
    const touch = e.touches[0];
    if (touch && touch.force !== undefined) {
      this._touchPressure.push(touch.force);
      if (this._touchPressure.length > 40) this._touchPressure.shift();
    }
  }

  _handleScroll() {
    const now    = Date.now();
    const scrollY = window.scrollY;
    if (this._lastScrollTs !== null) {
      const dt   = (now - this._lastScrollTs) / 1000;
      const rate = Math.abs(scrollY - this._lastScrollY) / Math.max(dt, 0.016);
      this._scrollRates.push(rate);
      if (this._scrollRates.length > 30) this._scrollRates.shift();
    }
    this._lastScrollY  = scrollY;
    this._lastScrollTs = now;
  }

  // ── On-device anomaly models ──────────────────────────────────────

  /** Coefficient of Variation of inter-key intervals.
   *  High CV → irregular typing → anomalous. */
  _keystrokeAnomaly() {
    const arr = this._keyIntervals;
    if (arr.length < 8) return 0.0;

    const mean = arr.reduce((a, b) => a + b, 0) / arr.length;
    if (mean === 0) return 0.0;

    const variance = arr.reduce((a, b) => a + (b - mean) ** 2, 0) / arr.length;
    const cv       = Math.sqrt(variance) / mean;   // 0 = perfectly uniform

    // Typical human typing CV ≈ 0.3–0.6. Bots/anomalies → near 0 or > 1.5
    const anomaly = cv < 0.1
      ? (0.1 - cv) / 0.1 * 0.8          // Suspiciously uniform (bot?)
      : cv > 0.8
        ? Math.min(1, (cv - 0.8) / 0.7)  // Suspiciously erratic
        : 0.0;

    return Math.round(anomaly * 100) / 100;
  }

  _touchAnomaly() {
    const arr = this._touchPressure;
    if (arr.length < 5) return 0.0;
    // Abnormally high or low pressure compared to baseline
    const avg = arr.reduce((a, b) => a + b, 0) / arr.length;
    return avg > 0.9 || avg < 0.05 ? 0.4 : 0.0;
  }
}


// ─────────────────────────────────────────────────────────────────────────────
//  React Native integration helper (for BOB Mobile App)
//  Usage: import { trustPulseMiddleware } from './trustpulse.js';
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Drop-in middleware for React Navigation's beforeRemove / beforeAction hooks.
 *
 * @example
 *   // In your Navigator
 *   const tp = new TrustPulse({ endpoint: '...' });
 *
 *   <Stack.Navigator
 *     screenOptions={{
 *       gestureEnabled: false,
 *     }}
 *   >
 *     <Stack.Screen
 *       name="Transfer"
 *       component={TransferScreen}
 *       listeners={trustPulseListener(tp, {
 *         eventType: 'transfer',
 *         onBlock: () => navigation.navigate('AccountLocked'),
 *         onStepUp: () => navigation.navigate('OTPVerification'),
 *       })}
 *     />
 *   </Stack.Navigator>
 */
export function trustPulseListener(tp, { eventType, amount, onBlock, onStepUp, onBiometric }) {
  return {
    beforeRemove: async (e) => {
      e.preventDefault();
      try {
        const result = await tp.score({ eventType, amount });
        switch (result.required_action) {
          case 'silent_pass':    e.data.action.payload && e.data.action(); break;
          case 'biometric_prompt': onBiometric?.(result); break;
          case 'step_up_auth':   onStepUp?.(result);   break;
          case 'block_and_alert': onBlock?.(result);   break;
        }
      } catch (err) {
        console.warn('[TrustPulse] Score failed, defaulting to step-up:', err);
        onStepUp?.({ error: true });
      }
    },
  };
}

/**
 * @typedef {object} ScoreResult
 * @property {string}   session_id
 * @property {number}   trust_score         0–100
 * @property {string}   risk_level          'low' | 'medium' | 'high' | 'critical'
 * @property {string}   required_action     'silent_pass' | 'biometric_prompt' | 'step_up_auth' | 'block_and_alert'
 * @property {number}   latency_ms
 * @property {string[]} explanation
 * @property {object}   breakdown
 * @property {string}   audit_ref
 */